docs/proposals/20210811-podmarker.md

```diff
@@ -0,0 +1,199 @@
+
+- Mark Pods that satisfy the specific probe (http/tcp/exec). We can add label based on query result so that we can rollback Pods based on the added labels.
+
+- Differentiate Pods with labels according to their different Nodes.
+
+## Proposal
+PodMarker is a new namespace-scoped CRD to add/remove labels/annotations of Pods.
+
+### User Stories
+
+#### Story 1: Mark Pods by LabelSelector
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: PodMarker
+metadata:
+  name: marker
+spec:
+  strategy:
+    replicas: 10
+    confilctPolicy: Overwrite
+  matchRequirements:
+    podSelector:
+      matchLabels:
+        app: nginx
+    nodeSelector:
+      matchLabels:
+        arch: x86
+  matchPreferences:
+  - podSelector:
+      matchLabels:
+        pod.state: working_online
+  - podReady: true
+  markItems:
+    labels:
+      upgradeStrategy: inPlace
+    annotations:
+      markUpgradeStrategyUsingPodMarker: marker
+```
+In this story, users may want to only update some Pods about nginx running in x86 arch machines, and they prefer to update the `Ready` Pods with label `"pod.state: working_online"`.
+
+For all Pods that already meet `.spec.matchRequirements`, PodMarker will mark these Pods orderly according to `spec.matchPreferences`:
+1. Firstly, mark the Pods with label `pod.state: working_online` **AND** `Ready`;
+2. Secondly, mark the Pods with label `pod.state: working_online`;
+3. Thirdly, mark the `Ready` Pods;
+4. Lastly, mark the other Pods.
+
+This PodMarker will help users mark these pods,
+and `spec.strategy` will ensure that at least 10 Pods are marked if there are at least 10 Pods meet `.spec.matchRequirements`.
+
+#### Story 2: Mark Pods by User-Defined Probe
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: PodMarker
+metadata:
+  name: marker
+spec:
+  strategy:
+    replicas: 50%
+    confilctPolicy: Overwrite
+  matchRequirements:
+    podSelector:
+      matchLabels:
+        app: nginx
+    nodeSelector:
+      matchLabels:
+        arch: x86
+    podProbes:
+    - containers:
+        - name: nginx
+        - name: sidecar
+      probe:
+        exec:
+          command:
+            - cat
+            - /tmp/healthy
+        initialDelaySeconds: 5
+        periodSeconds: 5
+      expectation: succeeded
+  markItems:
+    labels:
+      rollback: lastRevision
+    annotations:
+      markRollbackUsingPodMarker: marker
+```
+In this story, users define a probe to judge the health of `Pods`. This `PodMarker` will mark those `Pods` which satisfy all the following requirements:
+- having label "app: nginx"
+- running in the nodes with label "arch: x86"
+- probe result is "succeeded"
+****
+
+### Specification
+#### About Spec.Strategy
+- `Strategy.Replicas`:
+  - `Strategy.Replicas` can ensure that the number(or percentage) of marked Pods `>=` `Replicas`, unless:
```

Comment:
I have a question that why defines marked pods >= Replicas?
In canary release plan, we more likely need mark pods `<=` Replicas to  make sure the labeled pod  is updated done fist.But it also need cloneset support upgrade the pod with label matched, priority+partition can not ensure that.